torch/_inductor/codegen/cuda/gemm_template.py

```diff
@@ -1064,28 +1064,53 @@ def render(  # type: ignore[override]
-            )
-            read_names = OrderedSet(evt_read_names) - OrderedSet(evt_write_names)
-            write_names = OrderedSet(evt_write_names)
-            assert write_names, "There should be at least one write"
-            D_output_name = var_name_to_buffer_name["D"]
-            name_to_buffer = V.graph.name_to_buffer | V.graph.graph_inputs
-            D_output_buffer = name_to_buffer[D_output_name]
-            Y = D_output_buffer  # type: ignore[assignment]
-            # Interestingly, I don't think the rest of the layout matters here since we
-            # use the properties of the Y buffer to fill in D's properties in the epilogue
-            # args. This is needed though because it defines types expected in the epilogue args.
-            op.D.element = cutlass_utils.torch_dtype_to_cutlass_type(
-                D_output_buffer.get_layout().dtype
-            )
+        if epilogue_nodes or len(self.input_nodes) == 4:
+            if epilogue_nodes:
+                (
+                    evt_read_names,
+                    evt_write_names,
+                    var_name_to_buffer_name,
+                    evt_py_code,
+                ) = CutlassEVTCodegen.ir_to_evt_python_code(
+                    Y.get_name(), epilogue_nodes, V.kernel.removed_buffers
+                )
+                D_output_name = var_name_to_buffer_name["D"]
+                name_to_buffer = V.graph.name_to_buffer | V.graph.graph_inputs
+                D_output_buffer = name_to_buffer[D_output_name]
+                D_dtype = D_output_buffer.get_dtype()
+                Y = D_output_buffer  # type: ignore[assignment]
+                # Interestingly, I don't think the rest of the layout matters here since we
+                # use the properties of the Y buffer to fill in D's properties in the epilogue
+                # args. This is needed though because it defines types expected in the epilogue args.
+                op.D.element = cutlass_utils.torch_dtype_to_cutlass_type(
+                    D_output_buffer.get_layout().dtype
+                )
+
+                read_names = OrderedSet(evt_read_names) - OrderedSet(evt_write_names)
+                write_names = OrderedSet(evt_write_names)
+                assert write_names, "There should be at least one write"
+
+                input_names = list(read_names)
+                output_names = list(write_names)
+                epilogue_inputs = [name_to_buffer[name] for name in input_names]
+                epilogue_outputs = [name_to_buffer[name] for name in output_names]
+            else:  # Scaled MM, we read the two scale matrices and write a single output
+                (
+                    evt_read_names,
+                    var_name_to_buffer_name,
+                    evt_py_code,
+                ) = scaled_mm_evt(
+                    self.input_nodes[2].get_name(),
+                    self.input_nodes[3].get_name(),
+                    Y.get_name(),
+                )
```

Comment:
Nit: if this continues to grow, I think would make more sense to create and pass this function in on the callee side.

Reply:
I actually thought of a way to consolidate (and put this in the codegen module instead), but I'll do that in a separate PR